clang/lib/StaticAnalyzer/Checkers/BuiltinFunctionChecker.cpp

```diff
@@ -50,6 +101,44 @@ class BuiltinFunctionChecker : public Checker<eval::Call> {
 
 } // namespace
 
+void BuiltinFunctionChecker::HandleOverflowBuiltin(const CallEvent &Call,
+                                                   CheckerContext &C,
+                                                   BinaryOperator::Opcode Op,
+                                                   QualType ResultType) const {
+  // All __builtin_*_overflow functions take 3 argumets.
+  assert(Call.getNumArgs() == 3);
+
+  ProgramStateRef State = C.getState();
+  SValBuilder &SVB = C.getSValBuilder();
+  const Expr *CE = Call.getOriginExpr();
+
+  SVal Arg1 = Call.getArgSVal(0);
+  SVal Arg2 = Call.getArgSVal(1);
+
+  SVal RetVal = SVB.evalBinOp(State, Op, Arg1, Arg2, ResultType);
+
+  // TODO: Handle overflows with values that known to overflow. Like INT_MAX + 1
+  // should not produce state for non-overflow case and threat it as
```

Comment:
I didn't find a way to check this. We could do old-style thing like `res < 0 ? lhs > rhs : lhs < rhs` , but I guess, `SVB` should take care of it?

Reply:
I thought about this for some time and checked the implementation of `evalBinOp` (more precisely `SValBuilder:evalBinOpNN`, the case that's relevant here), but unfortunately I don't see a clear solution.

You are right that `SVB` _should_ take care of this, but unfortunately that does not happen currently.

Currently the analyzer (and `SValBuilder` in particular) has its own rigid opinion about overflow handling: it allows overflow on all numeric operations (including signed ones, where it's undefined behavior according to the standard) and models it without splitting the state, so e.g. if `x` and `y` are `signed char` values within the range `50 <= x, y <= 100`, then it calculates that the possible values of `x+y` are $[-128, -56] \cup [100, 127]$.

It would be good to enhance this behavior by tagging the ranges within the `RangeSet` with "did overflow happen?" bits that would let us perform a state split. This is theoretically doable (I don't see anything that would block this), but would be a complex refactoring effort (we'd need to pass along this "did we overflow?" bit through all the spaghetti code).

Without this, you can implement a workaround that calls `SValBuilder::getMinValue()` and `SValBuilder::getMaxValue()` to get the minimal and maximal values of the operands (as `llvm::APSInt` values, that is, **a**rbitrary **p**recision **s**igned **int**egers) and then uses these to determine which of the two branches (overflow, no-overflow) are possible.

This workaround solution would be better than the current "always act as if both branches were possible", but it is not sufficient for deducing the set of possible results on the no-overflow branch. (That is, if `x` and `y` are `signed char` values within the range `50 <= x, y <= 100`, then this workaround would create two branches: one where `__builtin_add_overflow` returns true, and and one where it returns `false` and stores a symbol with range $[-128, -56] \cup [100, 127]$ within the result pointer; while the correct assumption would be that the result has range $[100, 127]$ on this no-overflow branch.)

Reply:
evalBinOp models wrapping semantics - even for signed arithmetic (where it would be UB).
We don't have a similar evaluation model (operating on symbolic values, aka. SVals) which operates with mathematical (non-wrapping) semantics. This is unfortunate.

Initially I wanted to recommend `SVB.getKnownValue(State, Val)`, but I believe it's not too likely that one would use `malloc` with two concrete int params. So this wouldn't really enable many use-cases.

I can agree with @NagyDonat, about using `getMinValue` and friends. That would enable us using the constraints we have in the State, and unlock symbolic cases.

Reply:
So.... I've pushed what I came up with for handling overflow, but during test writing I found smth I don't understand. I've decided to push current state, since it's easier to show code than describe it =)

My current problem is following code:
```c
   if (a > 10)
     return;
   if (b > 10)
     return;

   // clang_analyzer_eval(a + b < 30); <--- Prints 1 and 0, but why ???
````
For some reason constraints do not work as expected. And because of that my overflow checker splits state where it shouldn't.... I'd really appreciate tips

Reply:
... which turns out to be bad idea, since it depends on `{i,u}128` support.

Will redo with Min/Max

Reply:
> `clang_analyzer_eval(a + b < 30); <--- Prints 1 and 0, but why ???`

Assuming that `a` and `b` are signed integers, they can be very negative, and then their sum can be a positive value above 30 (after an overflow). This means that both boolean values are possible for the expression `a + b < 30`, and the analyzer represents this by printing both 1 and 0.

(If I understand this correctly, we get two definite numbers instead of one range because the on-by-default `eagerlyAssume` mode causes a state split when it sees the comparison operator in `a + b < 30`, despite the fact that this is not in a conditional expression.)

Reply:
> Assuming that a and b are signed integers

Oh, sorry, I forgot to mention that `a` and `b` are unsigned. I've showed this in `test_uadd_overflow_contraints` test in this pr

Reply:
I did some digging in `RangedConstraintManager` and it seems that algebraic expressions that involve two symbolic values **are handled as black boxes** by the analyzer: in your example it creates a `SymSymExpr` that is known to be the sum of `a` and `b`, but we don't have any logic which would constrain the potential values of this `SymSymExpr` based on the constraints known about the operands. (The analyzer can propagate information in the easier case when it knows the exact value of one of the operands.)

I don't know why do we have this shameful limitation... I'd guess that implementing the necessary `RangeSet` operation wouldn't be prohibitively difficult, but there may be performance implications, and perhaps well-constrained (but not exactly known) values are not common enough to make this a valuable investment.

Reply:
I thought the ConstraintAssignor does the decomposition of the SymSymExpr and tracks the potential value range. It's probably somewhat limited though.